File: week2_algorithmic_warmup/4_least_common_multiple/lcm.cpp
#include <iostream>

/*
	Euclidean Algorithm to figure out the gcd
*/
int gcd_fast(int a, int b)
{
	// Ensure that a is not less than b
	if(a < b)
	{
		int temp(a);
		a = b;
		b = temp;
	}
	
	if(b == 0)
	{
		return a;
	}
	
	// change value of a to be the remainder of  a/b
	a = a % b;
	return gcd_fast(a, b);
}

long long lcm_naive(int a, int b) {
	// Linear Search
  /* for (long l = 1; l <= (long long) a * b; ++l) */
  /*   if (l % a == 0 && l % b == 0) */
  /*     return l; */

  /* return (long long) a * b; */

  /*
	For two numbers ‘a’ and ‘b’, LCM x GCD = a x b
  */
  // Step1:Find the Greatest Common Divisor of a and b
	int gcd = gcd_fast(a,b);
	
  // Step2: product
	long long prd = (long long) a * b;

	return (long long) prd / gcd;
}

int main() {
  int a, b;
  std::cin >> a >> b;
  std::cout << lcm_naive(a, b) << std::endl;
  return 0;
}

File: Algorithmic Warm Up/Least Common Multiple/lcm.py
# python3


def lcm_naive(a, b):
    assert 1 <= a <= 2 * 10 ** 9 and 1 <= b <= 2 * 10 ** 9

    multiple = max(a, b)
    while multiple % a != 0 or multiple % b != 0:
        multiple += 1

    return multiple

def gcd_fast(a, b):
    if a < b:
        temp = a
        a = b
        b = temp

    if b == 0:
        return a
    a = a % b
    return gcd_fast(a, b)

def lcm(a, b):
    assert 1 <= a <= 2 * 10 ** 9 and 1 <= b <= 2 * 10 ** 9

    gcd = gcd_fast(a, b)
    prd = a * b
    return prd / gcd


if __name__ == '__main__':
    input_a, input_b = map(int, input().split())
    print(lcm(input_a, input_b))
